src/FSharpPlus/DList.fs

```diff
@@ -1,45 +1,260 @@
-    let fold f x = List.fold f x << toList
-    let map f (DList x) = x [] |> List.map f |> List.append |> DList
-    let concat x = List.fold append empty x
-    let join (DList f) = concat (f [])
-    let ap f x = join <| map (fun y -> map ((|>) y) f) x
-    let bind m k = concat << List.map k << toList <| m
-
-open DList
+
+    //pattern discriminators  (active pattern)
+    let (|Cons|Nil|) (l : DList<'T>) = match l.TryUncons with Some(a,b) -> Cons(a,b) | None -> Nil
+
+    ///O(1). Returns a new DList of two lists.
+    let append left right = DList<'T>.appendLists(left, right)
+
+    ///O(1). Returns a new DList with the element added to the beginning.
+    let cons hd (l:DList<'T>) = 
+        match l.Length with
+        | 0 -> DList(1, Unit hd)
+        | _ -> DList(l.Length + 1, Join(Unit hd, l.dc) )
+
+    ///O(1). Returns DList of no elements.
+    [<GeneralizableValue>]
+    let empty<'T> : DList<'T> = DList(0, Nil )
+
+    ///O(n). Fold walks the DList using constant stack space. Implementation is from Norman Ramsey.
+    /// See http://stackoverflow.com/questions/5324623/functional-o1-append-and-on-iteration-from-first-element-list-data-structure/5334068#5334068
+    let foldBack (f : ('T -> 'State -> 'State)) (l:DList<'T>) (state : 'State) =
+        DList<'T>.foldBack f l state
+
+    let fold (f : ('State -> 'T -> 'State)) (state : 'State) (l:DList<'T>) =
+        DList<'T>.fold f state l
+
+    ///O(log n). Returns the first element.
+    let inline head (l:DList<'T>) = l.Head
+
+    ///O(log n). Returns option first element.
+    let inline tryHead (l:DList<'T>) = l.TryHead
+
+    ///O(1). Returns true if the DList has no elements.
+    let inline isEmpty (l:DList<'T>) = l.IsEmpty
+
+    ///O(1). Returns the count of elememts.
+    let inline length (l:DList<'T>) = l.Length
+
+    ///O(1). Returns DList of one elements.
+    let singleton x = DList(1, Unit x )
+
+    ///O(1). Returns a new DList with the element added to the end.
+    let inline conj x (l:DList<'T>) = l.Conj x
+
+    ///O(log n). Returns a new DList of the elements trailing the first element.
+    let inline tail (l:DList<'T>) = l.Tail
+
+    ///O(log n). Returns option DList of the elements trailing the first element.
+    let inline tryTail (l:DList<'T>) = l.TryTail
+
+    ///O(log n). Returns the first element and tail.
+    let inline uncons (l:DList<'T>) = l.Uncons
+
+    ///O(log n). Returns option first element and tail.
+    let inline tryUncons (l:DList<'T>) = l.TryUncons
+
+    ///O(n). Returns a DList of the seq.
+    let ofSeq s = DList<'T>.ofSeq s
+
+    ///O(n). Returns a list of the DList elements.
+    let inline toList l = foldBack (List.cons) l [] 
+
+    ///O(n). Returns a seq of the DList elements.
+    let inline toSeq (l:DList<'T>) = l :> seq<'T>
+
+    // additions to fit f#+ :
+    let inline map f (x:DList<_>)    = DList.foldBack (cons << f ) x empty
+    let concat x = DList.fold append empty x 
```

Comment:
Same here, it gets ``(DList<#DList<'b>> -> DList<'b>)``. Actually by removing the flexible type it leads to the same function as ``join``. In that case I would remove ``join`` since in most F# collections is it called ``concat``.

Reply:
I'll add that to the pull request.